### PR TITLE
[gateway] websocket transport

### PR DESCRIPTION
**Pull request by akarl10**
Microsoft now uses "plain" websocket as transport too.
At least two products are known: AVD/WVD and Windows Admin Center.

For the first case getting the required websocket url requires some extra step, for the windows admin center the url is
`https://<wac host>:<wac port>/api/nodes/<target hostname or ip>/features/tcp/3389`

since this commit is quite big here some notes for the review:
websocket.c contains the framing and unframing of websocket messages. It does not contain the http handshake.
because of performance reasons rdg.c still contains the outgoing websocket data framing.
the plain websocket transport is in wst.c. This one actually makes the http handshake. The interface used by transport.c is derived from rdg.c

I also already prepared a settings value for a http bearer auth that will be needed by azure gateway servers

### CODE REVIEW

**Comment by fifthdegree** [comment]
Just a comment: I'm not sure how useful it is to pass in a bearer token. We don't really expect a client to retrieve it manually. I understand it's for testing purposes, but unfortunately, the connection setup for avd is a bit more complex.

**Comment by akarl10** [comment]
> Just a comment: I'm not sure how useful it is to pass in a bearer token. We don't really expect a client to retrieve it manually. I understand it's for testing purposes, but unfortunately, the connection setup for avd is a bit more complex.

True, but I have this parameter mainly around for the purpose of being in the settings structure (like username and password). As you said the cli parameter is mainly for debugging (truth is that I have a unrelated websocket server impementation (not opensource, sorry) that has a bearer authentication and I tested the bearer auth against that before asking you to try against azure)

This can be changed, but my plan is the following:
after this gets merged I would suggest creating a new pseudo gateway transport arm.c that basically does
`POST https://<gateway in rdp file or cli>/api/arm/v2/connections` containing the required data (that is application id/ /app:||id and load-balance-info). If the bearer is available than use that, else do the same as you already do with /sec:aad (that is showing the web view or url to paste into the webbrowser)
as soon as the gateway url is fetched (and as far as I understood the ARRAffinity token)  it will be put into the settings structure and internally a fallback to the wst.c transport will be made (now having the bearer token in the settings structure thus not requiring fetching it again for the websocket connection)

**Comment by hardening** [comment]
@freerdp-bot test

**Comment by hardening** [comment]
Very interesting PR, thanks for the contribution.

**Comment by akallabeth** [comment]
@freerdp-bot test again please

File: libfreerdp/core/transport.h [hunk 1]
```diff
@@ -35,6 +35,7 @@ typedef enum
 
 #include "gateway/tsg.h"
 #include "gateway/rdg.h"
+#include "gateway/wst.h"
```
Comment (akallabeth) [inline libfreerdp/core/transport.h:38]: should be included in `transport.c`

File: libfreerdp/core/gateway/rdg.h [hunk 2]
```diff
@@ -36,6 +36,7 @@
 typedef struct rdp_rdg rdpRdg;
 
 #include "http.h"
+#include "websocket.h"
```
Comment (akallabeth) [inline libfreerdp/core/gateway/rdg.h:39]: should be included in `rdg.c`

File: libfreerdp/core/transport.c [hunk 3]
```diff
@@ -463,6 +464,29 @@ BOOL transport_connect(rdpTransport* transport, const char* hostname, UINT16 por
 
 	if (transport->GatewayEnabled)
 	{
+		if (!status && settings->GatewayUrl)
```
Comment (akallabeth) [inline libfreerdp/core/transport.c:467]: why check `!status` here?

File: libfreerdp/core/transport.c [hunk 4]
```diff
@@ -463,6 +464,29 @@ BOOL transport_connect(rdpTransport* transport, const char* hostname, UINT16 por
...
+		{
+			transport->wst = wst_new(context);
```
Comment (akallabeth) [inline libfreerdp/core/transport.c:469]: maybe `WINPR_ASSERT(!transport->wst);` to check for duplicate init

File: libfreerdp/core/gateway/wst.h [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+#include <freerdp/utils/ringbuffer.h>
+#include <freerdp/api.h>
+
+#include <freerdp/freerdp.h>
+#include <freerdp/types.h>
+#include <freerdp/settings.h>
+
+typedef struct rdp_wst rdpWst;
+
+#include "http.h"
```
Comment (akallabeth) [inline libfreerdp/core/gateway/wst.h:38]: best clean up includes, only include in header what is required for the function declarations/type definitions

File: libfreerdp/core/gateway/websocket.c [hunk 6]
```diff
@@ -0,0 +1,497 @@
...
+		BYTE data;
+		BYTE* partialMask = ((BYTE*)&maskingKey) + (streamPos % 4);
+		Stream_Read_UINT8(sPacket, data);
+		Stream_Write_UINT8(sWS, data ^ *partialMask);
+	}
+
+	Stream_SealLength(sWS);
+
+	ERR_clear_error();
+	status = BIO_write(bio, Stream_Buffer(sWS), Stream_Length(sWS));
```
Comment (akallabeth) [inline libfreerdp/core/gateway/websocket.c:90]: best `WINPR_ASSERT(bio);` here

File: libfreerdp/core/gateway/websocket.c [hunk 7]
```diff
@@ -0,0 +1,497 @@
...
+		/* biggest packet possible is 0xffff + 0xa, so 32bit is always enough */
+		Stream_Write_UINT32_BE(sWS, 0);
+		Stream_Write_UINT32_BE(sWS, payloadSize);
+	}
+	Stream_Write_UINT32(sWS, maskingKey);
+
+	/* mask as much as possible with 32bit access */
+	for (streamPos = 0; streamPos + 4 <= isize; streamPos += 4)
+	{
+		uint32_t masked = *((const uint32_t*)((const BYTE*)buf + streamPos)) ^ maskingKey;
```
Comment (akallabeth) [inline libfreerdp/core/gateway/websocket.c:147]: not sure this is endian safe? (there are people using it on such systems I´ve been told ;))

File: libfreerdp/core/gateway/websocket.c [hunk 8]
```diff
@@ -0,0 +1,497 @@
...
+	for (; streamPos < isize; streamPos++)
+	{
+		BYTE* partialMask = (BYTE*)(&maskingKey) + streamPos % 4;
+		BYTE masked = *((const BYTE*)((const BYTE*)buf + streamPos)) ^ *partialMask;
+		Stream_Write_UINT8(sWS, masked);
+	}
+
+	Stream_SealLength(sWS);
+
+	status = BIO_write(bio, Stream_Buffer(sWS), Stream_Length(sWS));
```
Comment (akallabeth) [inline libfreerdp/core/gateway/websocket.c:161]: best `WINPR_ASSERT(bio);` here

File: libfreerdp/core/gateway/websocket.c [hunk 9]
```diff
@@ -0,0 +1,497 @@
...
+	int status;
+
+	if (encodingContext->payloadLength == 0)
+	{
+		encodingContext->state = WebsocketStateOpcodeAndFin;
+		return 0;
+	}
+
+	ERR_clear_error();
+	status =
```
Comment (akallabeth) [inline libfreerdp/core/gateway/websocket.c:182]: best `WINPR_ASSERT(bio);` here

File: libfreerdp/core/gateway/websocket.c [hunk 10]
```diff
@@ -0,0 +1,497 @@
...
+
+	encodingContext->payloadLength -= status;
+
+	if (encodingContext->payloadLength == 0)
+		encodingContext->state = WebsocketStateOpcodeAndFin;
+
+	return status;
+}
+
+int websocket_read_wstream(BIO* bio, wStream* s, websocket_context* encodingContext)
```
Comment (akallabeth) [inline libfreerdp/core/gateway/websocket.c:220]: this function could be `static`?

File: libfreerdp/core/gateway/websocket.c [hunk 11]
```diff
@@ -0,0 +1,497 @@
...
+	if (Stream_GetRemainingCapacity(s) != encodingContext->payloadLength)
+	{
+		WLog_WARN(TAG,
+		          "wStream::capacity [%" PRIuz "] != encodingContext::paylaodLangth [%" PRIuz "]",
+		          Stream_GetRemainingCapacity(s), encodingContext->payloadLength);
+		return -1;
+	}
+
+	ERR_clear_error();
+	status = BIO_read(bio, Stream_Pointer(s), encodingContext->payloadLength);
```
Comment (akallabeth) [inline libfreerdp/core/gateway/websocket.c:243]: best `WINPR_ASSERT(bio);` here

File: libfreerdp/core/gateway/websocket.c [hunk 12]
```diff
@@ -0,0 +1,497 @@
...
+		closeDataLen = 2;
+
+	closeFrame = Stream_New(NULL, 6 + closeDataLen);
+	if (!closeFrame)
+		return FALSE;
+
+	Stream_Write_UINT8(closeFrame, WEBSOCKET_FIN_BIT | WebsocketPongOpcode);
+	Stream_Write_UINT8(closeFrame, closeDataLen | WEBSOCKET_MASK_BIT); /* no payload */
+	winpr_RAND((BYTE*)&maskingKey1, 2);
+	winpr_RAND((BYTE*)&maskingKey2, 2);
```
Comment (akallabeth) [inline libfreerdp/core/gateway/websocket.c:281]: `sizeof(maskingKey2`)`?

File: libfreerdp/core/gateway/websocket.c [hunk 13]
```diff
@@ -0,0 +1,497 @@
...
+		return FALSE;
+
+	Stream_Write_UINT8(closeFrame, WEBSOCKET_FIN_BIT | WebsocketPongOpcode);
+	Stream_Write_UINT8(closeFrame, 0 | WEBSOCKET_MASK_BIT); /* no payload */
+	winpr_RAND((BYTE*)&maskingKey, 4);
+	Stream_Write_UINT32(closeFrame, maskingKey); /* dummy masking key. */
+	Stream_SealLength(closeFrame);
+
+	ERR_clear_error();
+	status = BIO_write(bio, Stream_Buffer(closeFrame), Stream_Length(closeFrame));
```
Comment (akallabeth) [inline libfreerdp/core/gateway/websocket.c:324]: best `WINPR_ASSERT(bio);` here

File: libfreerdp/core/gateway/websocket.c [hunk 14]
```diff
@@ -0,0 +1,497 @@
...
+	if (status < 0)
+		return FALSE;
+	return TRUE;
+}
+
+static int websocket_handle_payload(BIO* bio, BYTE* pBuffer, size_t size,
+                                    websocket_context* encodingContext)
+{
+	int status;
+	BYTE effectiveOpcode = ((encodingContext->opcode & 0xf) == WebsocketContinuationOpcode
```
Comment (akallabeth) [inline libfreerdp/core/gateway/websocket.c:336]: best `WINPR_ASSERT(encodingContext);` here

File: libfreerdp/core/gateway/websocket.c [hunk 15]
```diff
@@ -0,0 +1,497 @@
...
+	/* return how many bytes have been written to pBuffer.
+	 * Only WebsocketBinaryOpcode writes into it and it returns directly */
+	return 0;
+}
+
+int websocket_read(BIO* bio, BYTE* pBuffer, size_t size, websocket_context* encodingContext)
+{
+	int status;
+	int effectiveDataLen = 0;
+	WINPR_ASSERT(encodingContext != NULL);
```
Comment (akallabeth) [inline libfreerdp/core/gateway/websocket.c:410]: best also `WINPR_ASSERT(bio);` here

File: libfreerdp/core/gateway/websocket.c [hunk 16]
```diff
@@ -0,0 +1,497 @@
...
+					encodingContext->fragmentOriginalOpcode = encodingContext->opcode;
+				encodingContext->state = WebsocketStateLengthAndMasking;
+			}
+			break;
+			case WebsocketStateLengthAndMasking:
+			{
+				BYTE buffer[1];
+				BYTE len;
+				ERR_clear_error();
+				status = BIO_read(bio, (char*)buffer, 1);
```
Comment (akallabeth) [inline libfreerdp/core/gateway/websocket.c:435]: `sizeof(buffer)`?

File: libfreerdp/core/gateway/websocket.c [hunk 17]
```diff
@@ -0,0 +1,497 @@
...
+			break;
+			case WebsocketStateShortLength:
+			case WebsocketStateLongLength:
+			{
+				BYTE buffer[1];
+				BYTE lenLength = (encodingContext->state == WebsocketStateShortLength ? 2 : 8);
+				while (encodingContext->lengthAndMaskPosition < lenLength)
+				{
+					ERR_clear_error();
+					status = BIO_read(bio, (char*)buffer, 1);
```
Comment (akallabeth) [inline libfreerdp/core/gateway/websocket.c:463]: `sizeof(buffer)`?

File: libfreerdp/core/gateway/websocket.c [hunk 18]
```diff
@@ -0,0 +1,497 @@
...
+	size_t len;
+	size_t fullLen;
+	int status;
+	wStream* sWS;
+
+	uint32_t maskingKey;
+
+	size_t streamPos;
+
+	len = Stream_Length(sPacket);
```
Comment (akallabeth) [inline libfreerdp/core/gateway/websocket.c:35]: `const size_t len`?

File: libfreerdp/core/gateway/websocket.c [hunk 19]
```diff
@@ -0,0 +1,497 @@
...
+	if (s != NULL)
+		return websocket_write_wstream(bio, s, WebsocketPongOpcode);
+
+	closeFrame = Stream_New(NULL, 6);
+	if (!closeFrame)
+		return FALSE;
+
+	Stream_Write_UINT8(closeFrame, WEBSOCKET_FIN_BIT | WebsocketPongOpcode);
+	Stream_Write_UINT8(closeFrame, 0 | WEBSOCKET_MASK_BIT); /* no payload */
+	winpr_RAND((BYTE*)&maskingKey, 4);
```
Comment (akallabeth) [inline libfreerdp/core/gateway/websocket.c:319]: `sizeof(maskingKey)`?

File: libfreerdp/core/gateway/wst.c [hunk 20]
```diff
@@ -0,0 +1,739 @@
...
+			return 0;
+	}
+
+	return nCount;
+}
+
+static int wst_bio_write(BIO* bio, const char* buf, int num)
+{
+	int status;
+	rdpWst* wst = (rdpWst*)BIO_get_data(bio);
```
Comment (akallabeth) [inline libfreerdp/core/gateway/wst.c:428]: `WINPR_ASSERT`?

File: libfreerdp/core/gateway/wst.c [hunk 21]
```diff
@@ -0,0 +1,739 @@
...
+	WINPR_UNUSED(bio);
+	WINPR_UNUSED(str);
+	WINPR_UNUSED(size);
+	return -2;
+}
+
+static long wst_bio_ctrl(BIO* bio, int cmd, long arg1, void* arg2)
+{
+	long status = -1;
+	rdpWst* wst = (rdpWst*)BIO_get_data(bio);
```
Comment (akallabeth) [inline libfreerdp/core/gateway/wst.c:504]: `WINPR_ASSERT`?

File: libfreerdp/core/gateway/wst.c [hunk 22]
```diff
@@ -0,0 +1,739 @@
...
+
+	return bio_methods;
+}
+
+static BOOL wst_parse_url(rdpWst* wst, const char* url)
+{
+	const char* hostStart;
+	const char* pos;
+
+	free(wst->gwhostname);
```
Comment (akallabeth) [inline libfreerdp/core/gateway/wst.c:604]: `WINPR_ASSERT`?

**Review by akallabeth — COMMENTED** [review]
Overall LGTM,

did add some (minor) remarks (mostly style/assert stuff)
nice addition!

just need a proper test setup for this, is there some howto to configure such a thing?

**Comment by akarl10** [comment]
@akallabeth wow, did not expect a review of such a big PR in less than a day.

I try to implement the changes you requested asap.

regarding the test setup:
https://learn.microsoft.com/en-us/windows-server/manage/windows-admin-center/deploy/install

you can install this even on windows 10/11. The downside is that it allows only the management of the host it runs on (but I would not be surprised if the rdp gateway component does not implement that restriction)

As said the avd/wvd thing still needs to be implemented, bud sadly I don't have a test setup for that.

About the question you asked about the endianess: the masking key is just a 4 byte blob. if the endianess of both masking key and data masking is the same it does not matter, it just needs to be consistent.

File: libfreerdp/core/gateway/websocket.c [hunk 23]
```diff
@@ -0,0 +1,497 @@
...
+static BOOL websocket_reply_pong(BIO* bio, wStream* s)
+{
+	wStream* closeFrame;
+	uint32_t maskingKey;
+	int status;
+
+	if (s != NULL)
+		return websocket_write_wstream(bio, s, WebsocketPongOpcode);
+
+	closeFrame = Stream_New(NULL, 6);
```
[the remark this answers is not in the document]
Comment (akarl10) [reply libfreerdp/core/gateway/websocket.c:313]: hmm. find it more readable if it is done the same way as everywhere. note that this frame will be sent just one time (if at all, the client might also initiate close the connection, so performance should not really matter)

File: libfreerdp/core/gateway/wst.h (hunk 5 above)
Comment (akarl10) [reply libfreerdp/core/gateway/wst.h]: I've chosen to include openssl/ssl.h for BIO*.
If you prefer can also change it to #include "../../crypto/tls.h"

File: libfreerdp/core/gateway/websocket.c (hunk 6 above)
Comment (akarl10) [reply libfreerdp/core/gateway/websocket.c:93]: I've put the asserts on the beginning of each function. There is no reason to try doing things if bio isn't there, so fail early

**Comment by akallabeth** [comment]
@freerdp-bot test

File: libfreerdp/core/gateway/wst.c [hunk 24]
```diff
@@ -0,0 +1,770 @@
...
+					response = http_response_recv(wst->tls, TRUE);
+					if (!response)
+					{
+						return FALSE;
+					}
+				}
+			}
+			StatusCode = http_response_get_status_code(response);
+			credssp_auth_free(wst->auth);
+			wst->auth = NULL;
```
Comment (akallabeth) [inline libfreerdp/core/gateway/wst.c:401]: missing `break;` (fallthrough produces warnings)

File: libfreerdp/core/gateway/wst.c [hunk 25]
```diff
@@ -0,0 +1,770 @@
...
+		char port[6];
+		pos++;
+		const char* portStart = pos;
+		while (*pos != '\0' && *pos != '/')
+			pos++;
+		if (pos - portStart > 5 || pos - portStart == 0)
+			return FALSE;
+		strncpy(port, portStart, (pos - portStart));
+		port[pos - portStart] = '\0';
+		int _p = atoi(port);
```
Comment (akallabeth) [inline libfreerdp/core/gateway/wst.c:671]: better use `strtol` (or one of the more fitting variants)

File: libfreerdp/core/gateway/websocket.c [hunk 26]
```diff
@@ -0,0 +1,521 @@
...
+	else if (len < 0x10000)
+		fullLen = len + 8; /* 2 byte "mini header" + 2 byte length + 4 byte masking key */
+	else
+		fullLen = len + 14; /* 2 byte "mini header" + 8 byte length + 4 byte masking key */
+
+	sWS = Stream_New(NULL, fullLen);
+	if (!sWS)
+		return FALSE;
+
+	winpr_RAND((BYTE*)&maskingKey, sizeof(maskingKey));
```
Comment (akallabeth) [inline libfreerdp/core/gateway/websocket.c:55]: no cast required

File: libfreerdp/core/gateway/websocket.c [hunk 27]
```diff
@@ -0,0 +1,521 @@
...
+	wStream* sWS;
+
+	uint32_t maskingKey;
+
+	int streamPos;
+
+	WINPR_ASSERT(bio);
+	WINPR_ASSERT(buf);
+
+	winpr_RAND((BYTE*)&maskingKey, sizeof(maskingKey));
```
Comment (akallabeth) [inline libfreerdp/core/gateway/websocket.c:116]: no cast required

File: libfreerdp/core/gateway/websocket.c [hunk 28]
```diff
@@ -0,0 +1,521 @@
...
+	if (s != NULL && Stream_Length(s) >= 2)
+		closeDataLen = 2;
+
+	closeFrame = Stream_New(NULL, 6 + closeDataLen);
+	if (!closeFrame)
+		return FALSE;
+
+	Stream_Write_UINT8(closeFrame, WEBSOCKET_FIN_BIT | WebsocketCloseOpcode);
+	Stream_Write_UINT8(closeFrame, closeDataLen | WEBSOCKET_MASK_BIT); /* no payload */
+	winpr_RAND((BYTE*)&maskingKey1, sizeof(maskingKey1));
```
Comment (akallabeth) [inline libfreerdp/core/gateway/websocket.c:294]: no cast required

File: libfreerdp/core/gateway/websocket.c [hunk 29]
```diff
@@ -0,0 +1,521 @@
...
+	winpr_RAND((BYTE*)&maskingKey2, sizeof(maskingKey2));
```
Comment (akallabeth) [inline libfreerdp/core/gateway/websocket.c:295]: no cast required

File: libfreerdp/core/gateway/websocket.c [hunk 30]
```diff
@@ -0,0 +1,521 @@
...
+	if (s != NULL)
+		return websocket_write_wstream(bio, s, WebsocketPongOpcode);
+
+	closeFrame = Stream_New(NULL, 6);
+	if (!closeFrame)
+		return FALSE;
+
+	Stream_Write_UINT8(closeFrame, WEBSOCKET_FIN_BIT | WebsocketPongOpcode);
+	Stream_Write_UINT8(closeFrame, 0 | WEBSOCKET_MASK_BIT); /* no payload */
+	winpr_RAND((BYTE*)&maskingKey, sizeof(maskingKey));
```
Comment (akallabeth) [inline libfreerdp/core/gateway/websocket.c:335]: no cast required

**Review by akallabeth — APPROVED** [review]
LGTM
some small style remarks left, but nothing that blocks.
(still need to do function testing though, but if anyone else already is doing that that is fine by me too)

**Comment by akallabeth** [comment]
@freerdp-bot test please

**Comment by akarl10** [comment]
just made a test with websockify (https://packages.debian.org/search?keywords=websockify)

I tested with
```bash
websokify --cert=/path/to/cert.pem --key=/path/to/key.pem 7777 targetserver:3389
```
you may create cert.pem and key.pem with
```
openssl req -x509 -newkey rsa:4096 -keyout key.pem -out cert.pem -sha256 -days 365 -nodes
```

then with this patch you can connect with
```
freerdp /v:targetserver /gateway:url:wss://websockify-hostname:7777/
```

**Comment by akallabeth** [comment]
@freerdp-bot test again please

**Comment by akallabeth** [comment]
@freerdp-bot test again please

**Comment by akallabeth** [comment]
@freerdp-bot test

**Comment by akallabeth** [comment]
@freerdp-bot come on, test please!

**Pull request merged**